src/browser/atom-application.coffee

```diff
@@ -428,12 +428,22 @@ class AtomApplication
 
   saveState: (allowEmpty=false) ->
     return if @quitting
+
+    Config = (require '../config')
+    config = new Config
+      configDirPath: process.env.ATOM_HOME
+      resourcePath: @resourcePath
+    config.setSchema null, {type: 'object', properties: _.clone(require('../config-schema'))}
+    config.load()
+    restorePreviousState = config.get('core.restorePreviousWindowsOnStart') ? true
+
     states = []
-    for window in @windows
-      unless window.isSpec
-        if loadSettings = window.getLoadSettings()
-          states.push(initialPaths: loadSettings.initialPaths)
-    if states.length > 0 or allowEmpty
+    if restorePreviousState
+      for window in @windows
+        unless window.isSpec
+          if loadSettings = window.getLoadSettings()
+            states.push(initialPaths: loadSettings.initialPaths)
+    if states.length > 0 or allowEmpty or not restorePreviousState
```

Comment:
I think it would be better to change the `loadState` method instead of `saveState`. With your current code, if there was a pre-existing `application.json` file, it would not be *updated*, but it would always be restored on startup. I think instead, you should make `loadState` return `false` immediately if `core.restorePreviousWindowsOnStart` is `false`.